File: kits/zig/test/namespace_test.zig
//! The kit's own names are names the notation cannot express.
//!
//! Article IV's identifier is a letter then letters and digits, so no
//! blueprint in any language can spell a name beginning with an underscore.
//! This kit asks a class for exactly one field of its own, the cell, and it is
//! named `_quo` for that reason.
//!
//! **Zig is the one kit where this defect is loud.** A struct may not carry
//! both a field and a decl of one name, so a cell named `quo` made a blueprint
//! declaring `quo()` impossible to write a class for — `error: duplicate
//! struct member name 'quo'`, at the being's own declaration. Loud, and still
//! a name the law allows that this kit could not serve.
//!
//! The handle side needs nothing: a field is reached by string through
//! `Handle.call`, and the handle's own facts are methods no field is written
//! over, so the two namespaces never touch here.

const std = @import("std");
const warden = @import("warden");
// Imported under a short name on purpose. A class declaring `quo()` shadows a
// module bound to `quo` inside its own scope, which is Zig's ordinary scoping
// and not this kit's business — but it is what an author of such a class meets,
// so the bench meets it too and names the way through.
const q = @import("quo");
const host = @import("host");

const Key = warden.Key;

fn still() i64 {
    return 1_000;
}

var draws: u64 = 0;

fn random() Key {
    draws += 1;
    var out: Key = undefined;
    var seed = draws *% 0x9e3779b97f4a7c15;
    for (&out) |*byte| {
        seed = seed *% 6364136223846793005 +% 1442695040888963407;
        byte.* = @truncate(seed >> 33);
    }
    return out;
}

/// Every name this kit has ever reached for on either side of the seam. On the
/// being side only `quo` was ever the kit's; the rest are the handle's own
/// methods, declared here to prove they were never the being's to lose.
const CLASH =
    \\Clash
    \\  quo() text
    \\  being() text
    \\  seal() text
    \\  send() text
    \\  describe() text
    \\  sketch() text
    \\  moved() text
    \\  blueprint() text
    \\  limit() text
    \\  text() text
    \\  digest() text
    \\  declares() text
    \\
;

const Clash = struct {
    _quo: q.Cell = .{},

    pub fn quo(_: *Clash) []const u8 {
        return "own:quo";
    }
    pub fn being(_: *Clash) []const u8 {
        return "own:being";
    }
    pub fn seal(_: *Clash) []const u8 {
        return "own:seal";
    }
    pub fn send(_: *Clash) []const u8 {
        return "own:send";
    }
    pub fn describe(_: *Clash) []const u8 {
        return "own:describe";
    }
    pub fn sketch(_: *Clash) []const u8 {
        return "own:sketch";
    }
    pub fn moved(_: *Clash) []const u8 {
        return "own:moved";
    }
    pub fn blueprint(_: *Clash) []const u8 {
        return "own:blueprint";
    }
    pub fn limit(_: *Clash) []const u8 {
        return "own:limit";
    }
    pub fn text(_: *Clash) []const u8 {
        return "own:text";
    }
    pub fn digest(_: *Clash) []const u8 {
        return "own:digest";
    }
    pub fn declares(_: *Clash) []const u8 {
        return "own:declares";
    }
};

const NAMES = [_][]const u8{
    "quo",   "being",     "seal",  "send", "describe", "sketch",
    "moved", "blueprint", "limit", "text", "digest",   "declares",
};

const World = struct {
    gpa: std.mem.Allocator,
    threaded: std.Io.Threaded = undefined,
    at_here: host.Host = undefined,
    at_there: host.Host = undefined,
    here: *warden.Warden = undefined,
    there: *warden.Warden = undefined,
    clash: Clash = .{},
    at_clash: q.Handle = undefined,

    fn init(gpa: std.mem.Allocator, self: *World) !void {
        self.* = .{ .gpa = gpa, .threaded = .init(gpa, .{}) };

        const io = self.threaded.io();
        inline for (.{ "at_here", "at_there" }) |which| {
            try host.Host.open(gpa, .{
                .seeds = host.seeds(random),
                .clock = still,
                .random = random,
                .io = io,
                .roads = &.{.memory},
            }, &@field(self, which));
        }
        self.here = &self.at_here.door;
        self.there = &self.at_there.door;
        self.at_clash = try q.holding(self.here, Clash, &self.clash, CLASH, .{}, gpa);
    }

    fn deinit(self: *World) void {
        self.at_here.close();
        self.at_there.close();
        self.threaded.deinit();
    }
};

test "every name this kit uses for itself is a field a being may declare" {
    const gpa = std.testing.allocator;
    var world: World = undefined;
    try World.init(gpa, &world);
    defer world.deinit();

    // A standing at it from the second ground, so the answers cross a door.
    const invitation = try world.at_clash.under.grant(gpa, world.at_clash.being());
    defer gpa.free(invitation.hints);
    const under = (q.Cell{ .door = world.there, .being = world.there.name }).at(gpa).?;
    var accepted = (try under.accept(invitation)).?;
    defer accepted.deinit();
    const far = accepted.only().?;

    // The value, never merely that nothing threw: two of the three ways this
    // defect showed itself in the other kits answered silence, which no caller
    // can tell from a refusal.
    inline for (NAMES) |name| {
        const said = (try far.call(gpa, []const u8, name, .{})) orelse
            return error.Silence;
        defer gpa.free(said);
        try std.testing.expectEqualStrings("own:" ++ name, said);
    }
}

test "the handle's own facts stand beside the fields, and answer their own" {
    const gpa = std.testing.allocator;
    var world: World = undefined;
    try World.init(gpa, &world);
    defer world.deinit();

    // `being` and `text` are the handle's, not the being's, and the class
    // declaring fields of those names takes nothing from them.
    try std.testing.expectEqual(world.at_clash.being(), world.clash._quo.being);
    try std.testing.expectEqualStrings(CLASH, world.at_clash.text());
    const said = (try world.at_clash.call(gpa, []const u8, "text", .{})).?;
    defer gpa.free(said);
    try std.testing.expectEqualStrings("own:text", said);
}

File: kits/js/bench/namespace.test.js
// The kit's own names are names the notation cannot express.
//
// Article IV's identifier is a letter then letters and digits, so no blueprint
// in any language can spell a name beginning with an underscore. This suite
// holds a being whose blueprint declares every name this kit ever used for
// itself and asserts each answers its own value through a handle — a remote
// one and a local one, because a handle keeps one shape wherever the being is.
//
// **It asserts the value, never that nothing threw.** Two of the three ways
// this defect showed itself were silent: a field the kit had written over
// answered `null`, which is indistinguishable from a refusal, a broken being
// or an absent one. A case that only checked for an exception would have
// passed against every one of them.
import test from 'node:test';
import assert from 'node:assert/strict';
import { Warden, memoryDelivery, seeds } from '../src/index.js';
import { hex } from '../src/bytes.js';

const still = () => 1_000;
const random = () => crypto.getRandomValues(new Uint8Array(32));

// Everything this kit has ever reached for on a being or a handle: the closure
// itself, the handle's own facts, the two halves of a seal, and the five looks.
// The list is the bench's, never the kit's — the kit guards no list at all.
const NAMES = [
  'quo',
  'being',
  'seal',
  'send',
  'describe',
  'sketch',
  'moved',
  'blueprint',
  'limit',
  'handles',
  'text',
  'digest',
  'declares',
];

const CLASH = `Clash\n${NAMES.map((name) => `  ${name}() text`).join('\n')}\n`;

class Clash {}
for (const name of NAMES) {
  Clash.prototype[name] = function () {
    return `own:${name}`;
  };
}

// A being that answers one of the two fields its blueprint declares. The
// unanswered one is what a being's own fault looks like from outside, and both
// handle shapes owe the caller the same silence for it.
const LOPSIDED = `Lopsided\n  here() text\n  absent() text\n`;
class Lopsided {
  here() {
    return 'here';
  }
}

const PEER = `Peer\n  poke() bool\n`;
class Peer {
  poke() {
    return true;
  }
}

async function world() {
  const delivery = memoryDelivery();
  const open = async (hint) => {
    const warden = await Warden.open({ seeds: seeds(random), clock: still, random, delivery });
    delivery.attach(hint, warden);
    warden.publish(hint);
    return warden;
  };
  const here = await open('mem://here');
  const there = await open('mem://there');

  const clash = new Clash();
  const peer = new Peer();
  const { handle: near } = await here.hold(clash, { blueprint: CLASH });
  await there.hold(peer, { blueprint: PEER });
  const [far] = await peer._quo.accept(await clash._quo.grant(clash), { label: 'clash' });
  return { here, there, clash, peer, near, far };
}

test('every name this kit uses for itself is a field a being may declare, across a door', async () => {
  const { far } = await world();
  for (const name of NAMES) {
    assert.equal(await far[name](), `own:${name}`, name);
  }
});

test('the same is true under one warden: one shape, and the same answers', async () => {
  const { near } = await world();
  for (const name of NAMES) {
    assert.equal(await near[name](), `own:${name}`, name);
  }
});

test("the kit's own machinery stands beside the fields it never ate", async () => {
  const { here, clash, far } = await world();
  // A handle that lost its own seal could not resend after silence, which is
  // the one safe act Article VIII gives a caller.
  assert.equal(hex(far._quo.being), hex(clash._quo.being));
  assert.equal(far._quo.text, CLASH);
  const sealed = await far._quo.seal('digest');
  assert.notEqual(sealed, null);
  assert.equal(await far._quo.send(sealed), 'own:digest');
  assert.notEqual(await far._quo.describe(), null);
  assert.equal(await far._quo.limit(), here.limit);
});

test('the closure is at _quo, so a being declaring quo() keeps its own method', async () => {
  const { clash, near } = await world();
  assert.equal(clash.quo(), 'own:quo');
  assert.equal(typeof clash._quo.grant, 'function');
  assert.equal(hex(near._quo.being), hex(clash._quo.being));
});

test("a being's own fault is the same silence at either shape", async () => {
  const { here, peer } = await world();
  const lopsided = new Lopsided();
  const { handle: near } = await here.hold(lopsided, { blueprint: LOPSIDED });
  const [far] = await peer._quo.accept(await lopsided._quo.grant(lopsided), { label: 'lopsided' });
  for (const handle of [near, far]) {
    assert.equal(await handle.here(), 'here');
    // Never a throw the far side would have swallowed.
    assert.equal(await handle.absent(), null);
  }
});

for (const reserved of ['cells', 'take']) {
  test(`a blueprint declaring ${reserved} is refused`, async () => {
    const { here } = await world();
    await assert.rejects(
      () => here.hold({}, { blueprint: `Thing\n  ${reserved}() bytes\n` }),
      /may not declare/,
      'the being provides this name, so a blueprint may not also claim it',
    );
  });
}

test('a name the blueprint does not declare is not on the handle at all', async () => {
  const { near, far } = await world();
  for (const handle of [near, far]) {
    assert.equal(handle.undeclared, undefined);
    assert.equal(handle.secret, undefined);
  }
});
